hcloud/util.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2018 Hetzner Cloud GmbH.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hcloud
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	robotmodels "github.com/syself/hrobot-go/models"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/hetznercloud/hcloud-cloud-controller-manager/internal/metrics"
+	"github.com/hetznercloud/hcloud-cloud-controller-manager/internal/robot"
+	"github.com/hetznercloud/hcloud-go/v2/hcloud"
+)
+
+func getCloudServerByName(ctx context.Context, c *hcloud.Client, name string) (*hcloud.Server, error) {
+	const op = "hcloud/getCloudServerByName"
+	metrics.OperationCalled.WithLabelValues(op).Inc()
+
+	server, _, err := c.Server.GetByName(ctx, name)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+
+	return server, nil
+}
+
+func getCloudServerByID(ctx context.Context, c *hcloud.Client, id int64) (*hcloud.Server, error) {
+	const op = "hcloud/getCloudServerByID"
+	metrics.OperationCalled.WithLabelValues(op).Inc()
+
+	server, _, err := c.Server.GetByID(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+	return server, nil
+}
+
+func getRobotServerByName(c robot.Client, node *corev1.Node) (server *robotmodels.Server, err error) {
+	const op = "hcloud/getRobotServerByName"
+
+	if c == nil {
+		return nil, errMissingRobotCredentials
+	}
+
+	// check for rate limit
+	if robot.IsRateLimitExceeded() {
+		return nil, fmt.Errorf("%s: rate limit exceeded - next try at %q", op, robot.TimeOfNextPossibleAPICall().String())
+	}
```

Comment:
What is the recommended behavior ? Should we fail early or block until the limit is down ? I assume this is part of a reconciliation loop, so maybe failing early is best ? What happens if we stop in the middle of the reconciliation and the state is only half applied ?

Reply:
Yes, all of this is called in a reconciliation loop so failing early is preferred.

If we stop in the middle, only the state until then is applied and everything else will get applied with the next loop. This is not really a problem here, because the reconciliation loop is ran for each node separately.